distribution/docker/build.gradle

```diff
@@ -0,0 +1,122 @@
+  dockerPlugins project(path: ":plugins:ingest-user-agent", configuration: 'zip')
+  dockerSource project(path: ":distribution:archives:tar")
+  ossDockerSource project(path: ":distribution:archives:oss-tar")
+}
+
+ext.expansions = { oss ->
+  return [
+    'elasticsearch' : oss ? "elasticsearch-oss-${VersionProperties.elasticsearch}.tar.gz" : "elasticsearch-${VersionProperties.elasticsearch}.tar.gz",      
+    'jdkUrl' : 'https://download.java.net/java/GA/jdk11/13/GPL/openjdk-11.0.1_linux-x64_bin.tar.gz',
+    'jdkVersion' : '11.0.1',
+    'license': oss ? 'Apache-2.0' : 'Elastic License',
+    'ingest-geoip' : "ingest-geoip-${VersionProperties.elasticsearch}.zip",
+    'ingest-user-agent' : "ingest-user-agent-${VersionProperties.elasticsearch}.zip",
+    'version' : VersionProperties.elasticsearch
+  ]
+}
+
+static String files(final boolean oss) {
+  return "build/${ oss ? 'oss-' : ''}docker"
+}
+
+static String taskName(final String prefix, final boolean oss, final String suffix) {
+  return "${prefix}${oss ? 'Oss' : ''}${suffix}"
+}
+
+void addProcessDockerfileTask(final boolean oss) {
+  task(taskName("process", oss, "Dockerfile"), type: Copy) {
+    into files(oss)
+
+    from 'src/docker/Dockerfile'
+    MavenFilteringHack.filter(it, expansions(oss))
+  }
+}
+
+void addCopyBinFilesTask(final boolean oss) {
+  task(taskName("copy", oss, "BinFiles"), type: Copy) {
+    into "${files(oss)}/bin"
+
+    from 'src/docker/bin'
+  }
+}
+
+void addCopyConfigFilesTask(final boolean oss) {
+  task(taskName("copy", oss, "ConfigFiles"), type: Copy) {
+    into files(oss)
+
+    from 'src/docker/config'
+  }
+}
+
+void addCopyElasticsearchDistributionTask(final boolean oss) {
+  task(taskName("copy", oss, "ElasticsearchDistribution"), type: Copy) {
+    into files(oss)
+
+    if (oss) {
+      from configurations.ossDockerSource
+    } else {
+      from configurations.dockerSource
+    }
+  }
+}
+
+void addCopyPluginsTask(final boolean oss) {
+  task(taskName("copy", oss, "Plugins"), type: Copy) {
+    into files(oss)
+
+    from configurations.dockerPlugins
+  }
+}
+
+void addBuildDockerImage(final boolean oss) {
+  task(taskName("build", oss, "DockerImage"), type: LoggedExec) {
+    dependsOn taskName("process", oss, "Dockerfile")
+    dependsOn taskName("copy", oss, "BinFiles")
+    dependsOn taskName("copy", oss, "ConfigFiles")
+    dependsOn taskName("copy", oss, "ElasticsearchDistribution")
+    dependsOn taskName("copy", oss, "Plugins")
+    final List<String> tags
+    if (oss) {
+      tags = [ "docker.elastic.co/elasticsearch/elasticsearch-oss:${VersionProperties.elasticsearch}" ]
+    } else {
+      tags = [
+        "elasticsearch:${VersionProperties.elasticsearch}",
+        "docker.elastic.co/elasticsearch/elasticsearch:${VersionProperties.elasticsearch}",
+        "docker.elastic.co/elasticsearch/elasticsearch-full:${VersionProperties.elasticsearch}"
+      ]
+    }
+    executable 'docker'
```

Comment:
This means that we'll need docker installed even for `./gradlew assemble`. 
Are we ok with any version of docker for that? 
It  might make sense to add up-front checks like we have for the JDK to make sure it's there and it works.  
Permissions are also something we are likely to run into as many tutorials online advertise the use of `sudo docker` people might not have permissions set up for the user running the build.

Reply:
@atorok I pushed 684fcb5d6f3a60c1a05dc1d01f8b2392e4984dc8.

Reply:
@atorok I pushed one more change, to check if we can even run Docker: 26fe6061e2055660f21d7b033dfee7237be908b3

This gives output like:

```
FAILURE: Build failed with an exception.

* What went wrong:
a problem occurred running Docker yet it is required to run the following tasks: 
  :distribution:docker:buildDockerImage
  :distribution:docker:buildOssDockerImage
the problem is that [docker images] exited with exit code [1] with error output [Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Get http://%2Fvar%2Frun%2Fdocker.sock/v1.39/images/json: dial unix /var/run/docker.sock: connect: permission denied]
```

Your input @nik9000 @rjernst would be appreciated here too!

Reply:
That seems right. We talked about making a parameter that let you explicitly skip the tasks which would be a nice thing to have. Something like `-Dbuild.docker=false`.

Reply:
@nik9000 I pushed b20be22043058eeedf5e09a6a47a450589233713. It's okay?

Reply:
I think this would be an ideal time to consider using [img](https://github.com/genuinetools/img) or [Buildah](https://github.com/containers/buildah) as the build executor instead of the Docker daemon. I personally consider this to be the "next step" in improving the build, and it just seems like a great time to redefine the dependencies.

Getting away from the Docker daemon not only improves the local build experience for developers, especially if they don't run Linux on their workstations, it also makes a much better story for running builds in container environments, like the work that @mgreau is actively pursuing.

Reply:
Thank you for the suggestion. Alas the Docker daemon will stay around since we are migrating some of our testing away from Vagrant fixtures to Docker Compose fixtures. That’s not to say we won’t take your suggestion up.